### DISCUSSION TITLE
How can we run an xctestrun with go-ios?

### DISCUSSION THREAD

**Discussion by sepsky** [danielpaulus/go-ios#253] [Q&A]
Hi, 

I am using go-ios to deploy and test my app. 
For this, I've already created the IPA for the main app, as well as the runner, and an xctestrun. So I have the following already: 
- My app, called XYZ.app
- My Runner, called XYZ-Runner.app
- My xctestrun, called schemeName_iphoneos16.2-arm64.xctestrun

I've installed the two apps on my device using go-ios, and they're both there. I then ran the following command (AFTER trusting the apps / certs / authorities in the Settings menu of the device):
```
ios runtest --bundle-id=BUNDLE_ID_OF_THE_APP --test-runner-bundle-id= BUNDLE_ID_OF_THE_RUNNER_APP
```

I get this error:
```
{"error":"write unix -\u003e/var/run/usbmuxd: write: broken pipe","level":"info","msg":"Failed running Xcuitest","time":"2023-04-27T14:21:25+01:00"}
```
One observation, because I'm running the runner app, I don't have the xcconfig to pass to `go-ios`. Is there something I can do for this, and do you think this is the reason why it's failing? 

Thanks a lot and if I can provide any further information, please let me know. 

Best wishes

Sep

**Comment by wellkilo** [comment]
Since you already have the .xctestrun file, use the command designed for it instead of reconstructing bundle/config arguments manually:

    ios runxctest \
      --xctestrun-file-path /path/to/schemeName_iphoneos16.2-arm64.xctestrun \
      --log-output test.log \
      --junit-output results.xml \
      --udid YOUR_UDID

The .xctestrun contains TestHostBundleIdentifier, TestBundlePath, environment variables, OnlyTestIdentifiers/SkipTestIdentifiers and whether the target is a UI test. Current go-ios parses format versions 1 and 2 and builds TestConfig from them: [xctestrunutils.go](https://github.com/danielpaulus/go-ios/blob/c584042907dee35bf80377284ae3aac85990b4d7/ios/testmanagerd/xctestrunutils.go#L17-L24).

For ios runtest, --xctest-config means the installed .xctest bundle name (for example XYZUITests.xctest), not an xcconfig build-settings file. Omitting/deriving that incorrectly can prevent the bundle from loading.

The usbmux broken pipe may also reflect the old go-ios/iOS protocol path used in 2023. Re-test with a current binary and healthy usbmuxd; for iOS 17+ run sudo ios tunnel start first. The current CLI exposes runxctest plus log/JUnit output directly: [main.go](https://github.com/danielpaulus/go-ios/blob/c584042907dee35bf80377284ae3aac85990b4d7/main.go#L478-L482).
